### PR TITLE
Feature/230 efo polling refactor

### PR DESCRIPTION
**Pull request by Elaine-Krauss-TCG**
[Fecfile#230](https://fecgov.atlassian.net/browse/FECFILE-230)

### CODE REVIEW

**Review by dheitzer — CHANGES_REQUESTED** [review]
Looks nice! only things is - are the changes to [performance-testing/bulk-testing-data-fixture-generator.py](https://github.com/fecgov/fecfile-web-api/pull/1083/files#diff-e7d1f1ce5369d88d0a406247d6e408421e59d3b0656e869ba3ef5e3a95a9c879) for another ticket? they seem unrelated. 

Also, it might be worth putting in the original ticket (230) a comment about what requirements were determined with FEC relating to this ticket.

**Review by toddlees — DISMISSED** [review]

**Comment by Elaine-Krauss-TCG** [comment]
> Looks nice! only things is - are the changes to [performance-testing/bulk-testing-data-fixture-generator.py](https://github.com/fecgov/fecfile-web-api/pull/1083/files#diff-e7d1f1ce5369d88d0a406247d6e408421e59d3b0656e869ba3ef5e3a95a9c879) for another ticket? they seem unrelated.

Good question.  In order to generate a sufficiently large dotFEC file within a reasonable timeframe, I needed to add memo texts to every transaction.  I added additional functionality to this script in order to make that process quicker.

File: django-backend/fecfiler/settings/base.py [hunk 1]
```diff
@@ -352,6 +354,22 @@ def get_logging_processors():
...
+"""EFO POLLING SETTINGS
+"""
+EFO_POLLING_MAX_DURATION_DEFAULT = 300  # Max duration of polling in seconds
+EFO_POLLING_INTERVAL_DEFAULT = 30  # Time between requests in seconds
+
+EFO_POLLING_MAX_DURATION = get_float_from_string(
+    env.get_credential("EFO_POLLING_MAX_DURATION"),
+    EFO_POLLING_MAX_DURATION_DEFAULT
+)
+EFO_POLLING_INTERVAL = get_float_from_string(
```
[the remark this answers is not in the document]
Comment (Elaine-Krauss-TCG) [reply django-backend/fecfiler/settings/base.py:367]: The idea here is that this is the fallback for converting the string value of the environment variable into a float.  Given the next request, though, this is unnecessary.

File: django-backend/fecfiler/shared/utilities.py [hunk 2]
```diff
@@ -14,3 +14,13 @@ def get_model_data(data, model):
...
     return {field: data[field] for field in field_names if field in data}
+
+
+def get_float_from_string(string, fallback):
+    try:
+        value = float(string)
+        if value is not None:
+            return value
+        return fallback
+    except Exception:
```
[the remark this answers is not in the document]
Comment (Elaine-Krauss-TCG) [reply django-backend/fecfiler/shared/utilities.py:25]: Changed!  It now raises an error when the server spins up with a provided value for these environment variables that cannot be converted to a float.

image

File: django-backend/fecfiler/web_services/tasks.py [hunk 3]
```diff
@@ -154,30 +195,75 @@ def submit_to_webprint(
...
+            submission.fec_batch_id, submission.fec_submission_id
+        )
+        submission.save_fec_response(status_response_string)
+        if (
+            submission.fec_status in FECStatus.get_terminal_statuses_strings()
+            or submission.fecfile_polling_attempts >= EFO_POLLING_MAX_ATTEMPTS
+        ):
+            if submission.fecfile_polling_attempts >= EFO_POLLING_MAX_ATTEMPTS:
+                logger.warning("POLLING ATTEMPTS EXCEEDED")
+                submission.fec_status = FECStatus.FAILED
```
[the remark this answers is not in the document]
Comment (Elaine-Krauss-TCG) [reply django-backend/fecfiler/web_services/tasks.py:250]: Great idea :+1:

File: django-backend/fecfiler/web_services/test_tasks.py [hunk 4]
```diff
@@ -184,3 +196,119 @@ def test_submit_to_webprint(self):
         self.assertEqual(
             webprint_submission.fec_image_url, "https://www.fec.gov/static/img/seal.svg"
         )
+
+
+class TestWebPrintSubmitter(WebPrintSubmitter):
```
[the remark this answers is not in the document]
Comment (Elaine-Krauss-TCG) [reply django-backend/fecfiler/web_services/test_tasks.py:201]: I changed it to UnitTestWebPrintSubmitter which is at least a little bit better.  `AlwaysProcessingWebPrintSubmitter` would be great, except the behavior here is _very_ specific to what's needed for testing.  I've added a comment to explain the behavior.

**Comment by Elaine-Krauss-TCG** [comment]
> Also, it might be worth putting in the original ticket (230) a comment about what requirements were determined with FEC relating to this ticket.

@dheitzer We determined in a prior weekly dev meeting that, for this ticket, we'll set things up so that the values can be configured with environment variables, and then we'll get the specifics of the requirements in the future.  As long as that's still good with @lbeaufort then we _should_ be good.

**Comment by lbeaufort** [comment]
> @dheitzer We determined in a prior weekly dev meeting that, for this ticket, we'll set things up so that the values can be configured with environment variables, and then we'll get the specifics of the requirements in the future. As long as that's still good with @lbeaufort then we _should_ be good.

@Elaine-Krauss-TCG I think since this dev work is complete, we should move it forward/merge the changes and can tweak the config once we get additional requirements.

**Review by dheitzer — APPROVED** [review]

**Review by sasha-dresden — APPROVED** [review]

**Pull request merged**